rules/RULE_10_1_A_do_not_use_bufferoverflow_risky_function_for_unix.py

```diff
@@ -42,6 +42,13 @@ def RunRule(lexer, contextStack):
             t2 = lexer.PeekNextTokenSkipWhiteSpaceAndComment()
             if t2 is not None and t2.type == "LPAREN":
                 t3 = lexer.PeekPrevTokenSkipWhiteSpaceAndComment()
+                # if :: comes first then we are not using dangerous C
+                # functions, so we can get out
+                if t3.type == "DOUBLECOLON":
+                    # check out the namespace that gives us this ::
+                    prev_namespace_token = lexer.GetPrevTokenInType("ID", keepCur=True)
+                    if prev_namespace_token.value != "std":
+                        return

```

Comment:
I don't like the fact that all you have to do to hide from the error is put the function in a namespace.

Would you be more amenable to a warning saying something like "Function bears a name long associated with dangerous bugs. Just make sure your namespace isn't giving you similar headaches in future" (or something better :smile: ) and making another carve-out for "reputed libraries" that don't suffer from the problem (such as "fmt")?

Reply:
I'm not sure I understand what you are saying: if you use the fmt library then you are not calling the stdio's sprintf() at all: you are calling a safe modern C++ function that uses strings.

But what you say relates to what I think the real problem with my patch is: all you have to do to confuse my logic is to rename std to something else.  For example, this would pass:
```
#include <cstdio>

namespace dude = std;

int main()
{
  char s[10];
  dude::sprintf(s, "this is a string longer than 10");
  printf("%s\n", s);
}
```
Although interestingly a modern compiler will track what's done with sprintf() and warn on this, and a modern run time will alert you to a smashed stack.

I don't have a solution to that (rather unlikely) problem, but since these are all heuristic measures anyway, I think mine is much closer to always doing the right thing.

And I agree with you that a good solid warning would be better than silently agreeing.  We require a clean pass of nsiqcppstyle, so I wonder if it makes sense to have warnings.  I haven't thought about that much.

Reply:
> We require a clean pass of nsiqcppstyle

That shouldn't be an issue.

Maybe, the following rephrasing of the solution will help:

* For all sprintf and cousins, no namespace => error
* If namespace is present and in blacklist (std) => error
* If namespace is present and in whitelist (fmt) => no error
* Else => Warning unknown safety. Take manual actions to rectify situation

Reply:
I see; I'll look at doing something that meets your spec.

Reply:
Any updates on this?

Reply:
If there's' no progress on this for a while, I'll proceed on my own. This is a nice PR and I don't want it to go unmerged.